pre-cache/pull

```diff
@@ -27,59 +42,88 @@ mirror_images() {
             current_pull=$((current_pull + 1))
             continue
         fi
-        $container_tool pull $img --authfile=/var/lib/kubelet/config.json -q > /dev/null &
+        $container_tool pull $img --authfile=$pull_secret_path -q > /dev/null &
         #$container_tool copy docker://${img} --authfile=/var/lib/kubelet/config.json containers-storage:${img} -q & # SKOPEO 
         pids[${img}]=$! # Keeping track of the PID and container image in case the pull fails
         max_bg=$((max_bg - 1)) # Batch size adapted 
         current_pull=$((current_pull + 1)) 
         if [[ $max_bg == 0 ]] # If the batch is done, then monitor the status of all pulls before moving to the next batch
         then
           for pid in ${!pids[@]}; do
-            wait ${pids[$pid]} # The way wait monitor for each background task (PID). If any error then copy the image in the failed array so it can be retried later
-            if [[ $? != 0 ]]; then
-              log_debug "Pull failed for container image: ${pid} . Retrying later... "
-              failed_pulls+=(${pid}) # Failed, then add the image to be retrieved later
-            fi
+            wait_image $pid ${pids[$pid]}
           done
           # Once the batch is processed, reset the new batch size and clear the processes hash for the next one
           max_bg=$max_pull_threads
           pids=()
         fi
     done
+
+    # Make sure that all pull-threads have been processed
```

Comment:
Why do we need to redo this?

Reply:
This is required especially for cases when the number of images to be pulled are less than the batch size (`max_bg`). Without the added logic, assuming the pulls commence without any failures, the script proceeds normally to check for failed pulls and then terminates if no pulls have failed returning a success. With termination of the script - the Kubernetes job also terminates - resulting in termination of the background pull threads. If the image size is large - the pull thread will not finish in time (i.e. end of script) thus pre-caching will pre-emptively fail and incorrectly report a success. I observed this with manual tests involving 3 additional-images, 1 of which was close to a Gigabyte.
Thus, it is necessary to introduce a wait on background threads before proceeding to retrying failed images.

Reply:
Also worthy to note that when the number of image to be pulled < `max_bg`, then the first `wait_image` loop is not executed.